### PR TITLE
Support pickling to CaffeFunction

### LINKED ISSUE #297: CaffeFunction cannot be pickled
Some attributes (`forwards`, `layers` and `variables`) of CaffeFunction cannot be pickled directly. We have to provide a customized pickling methods (e.g. `__getstate__` and `__setstate__`).

(This bug is reported by [jellied_unagi](https://twitter.com/jellied_unagi) on Twitter. Thank you!)

### PR DESCRIPTION
**Pull request by okuta**
fix #297.
If you want to test , please insert the following code to the 58-th row of modelzoo example.

```python:
pickle.dump(func, open("model.pickle", 'wb'), -1)
func = pickle.load(open("model.pickle", 'rb'))
```

https://github.com/pfnet/chainer/blob/v1.3.0/examples/modelzoo/evaluate_caffe_net.py#L58

### CODE REVIEW

File: chainer/functions/caffe/caffe_function.py [hunk 1]
```diff
@@ -202,7 +202,8 @@ def _add_layer(self, layer):
         bottom = []
         for blob_name in layer.bottom:
             bottom.append(self.split_map.get(blob_name, blob_name))
-        self.layers.append((layer.name, bottom, layer.top))
+        top = [i for i in layer.top]
```
Comment (unnonouno) [inline chainer/functions/caffe/caffe_function.py]: `list(layer.top)` ?

**Comment by okuta** [comment]
Thank you for your suggestion.

File: chainer/functions/caffe/caffe_function.py [hunk 2]
```diff
@@ -399,3 +400,34 @@ def _get_width(blob):
...
+        self.func = func
+        self.kwargs = kwargs
+
+    def __call__(self, *xs):
+        return self.func(xs, **self.kwargs)
+
+
+class _DropoutFunction(object):
+    def __init__(self, caffe_func, ratio):
+        self.caffe_func = caffe_func
```
Comment (unnonouno) [inline chainer/functions/caffe/caffe_function.py]: Why don't you use `self.train = train`?

Comment (unnonouno) [inline chainer/functions/caffe/caffe_function.py]: OK, I understood it. Please write comment about it.

**Comment by okuta** [comment]
I wrote comment.

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**